alternatives.c

```diff
@@ -321,96 +335,102 @@ static int readConfig(struct alternativeSet *set, const char *title,
             fprintf(stderr, _("path %s unexpected in %s\n"), line, path);
-            free(line);
-            return 1;
+            r = 1;
+            goto finish;
         }
 
         groups = realloc(groups, sizeof(*groups) * (numGroups + 1));
-        groups[numGroups].title = line;
+        groups[numGroups].title = strsteal(&line);
 
-        line = parseLine(&buf);
+        nextLine(&buf, &line);
         if (!line || !*line) {
             fprintf(stderr, _("missing path for follower %s in %s\n"), line, path);
-            return 1;
+            r = 1;
+            goto finish;
         }
 
-        groups[numGroups++].facility = line;
+        groups[numGroups++].facility = strsteal(&line);
 
-        line = parseLine(&buf);
+        nextLine(&buf, &line);
     }
 
     if (!line) {
         fprintf(stderr, _("unexpected end of file in %s\n"), path);
-        return 1;
+        r = 1;
+        goto finish;
     }
 
-    line = parseLine(&buf);
+    nextLine(&buf, &line);
     while (line && *line) {
         set->alts = realloc(set->alts, (set->numAlts + 1) * sizeof(*set->alts));
 
         if (*line != '/') {
             fprintf(stderr, _("path to alternate expected in %s\n"), path);
             fprintf(stderr, _("unexpected line in %s: %s\n"), path, line);
-            return 1;
+            r = 1;
+            goto finish;
         }
 
         set->alts[set->numAlts].leader.facility = strdup(normalize_path(groups[0].facility));
         set->alts[set->numAlts].leader.title = strdup(groups[0].title);
-        set->alts[set->numAlts].leader.target = line;
+        set->alts[set->numAlts].leader.target = strsteal(&line);
         set->alts[set->numAlts].numFollowers = numGroups - 1;
         if (numGroups > 1)
             set->alts[set->numAlts].followers = malloc(
                 (numGroups - 1) * sizeof(*set->alts[set->numAlts].followers));
         else
             set->alts[set->numAlts].followers = NULL;
 
-        line = parseLine(&buf);
         set->alts[set->numAlts].priority = -1;
         set->alts[set->numAlts].initscript = NULL;
         set->alts[set->numAlts].family = NULL;
 
-        if (line && line[0] == '@') {
-            line++;
-            end = strchr(line, '@');
-            if (!end || (end == line)) {
+        nextLine(&buf, &line);
+        ptr = line;
+
+        if (ptr && ptr[0] == '@') {
+            ptr++;
+            end = strchr(ptr, '@');
+            if (!end || (end == ptr)) {
                 fprintf(stderr,
                         _("closing '@' missing or the family is empty in %s\n"),
                         path);
                 fprintf(stderr, _("unexpected line in %s: %s\n"), path, line);
-                return 1;
+                r = 1;
+                goto finish;
             }
             *end = '\0';
-            set->alts[set->numAlts].family = strdup(line);
-            line = end + 1;
+            set->alts[set->numAlts].family = strdup(ptr);
```

Comment:
Ditto. `set->numAlts` is only incremented at the end. If `goto finish` happens before that, the current, partially initialized, entry won't be freed.